Make data streams a basic licensed feature.

martijnvg opened the pull request:
* Create new data-stream xpack module.
* Move TimestampFieldMapper to the new module,
  this results in storing a composable index template
  with data stream definition only to work with default
  distribution. This way data streams can only be used
  with default distribution, since a data stream can
  currently only be created if a matching composable index
  template exists with a data stream definition.
* Add `validate()` method to `MetadataFieldMapper` so that
   in core the validation that exists in `TimestampFieldMapper`
   can be invoked.
* Move data stream integration tests to xpack module.

In a follow up the data stream transport and rest actions can be moved to the xpack data-stream module.

martijnvg committed:
Make data streams a basic licensed feature.

* Create new data-stream xpack module.
* Move TimestampFieldMapper to the new module,
  this results in storing a composable index template
  with data stream definition only to work with default
  distribution. This way data streams can only be used
  with default distribution, since a data stream can
  currently only be created if a matching composable index
  template exists with a data stream definition.

In a follow up the data stream transport and
rest actions can be moved to the xpack data-stream module.

martijnvg committed:
move more data stream yaml tests

martijnvg committed:
copied resolve index yaml to data stream module and modify the existing to not use data streams

martijnvg committed:
moved more integ tests to xpack module

martijnvg committed:
Added test that verifies behaviour if composable index template is used with data steam definition and it isn't supported.

martijnvg committed:
fixed checkstyle issues

martijnvg committed:
Remove interface and move validate() method to MetadataFieldMapper class

martijnvg committed:
fixed test

martijnvg committed:
added jdocs

martijnvg committed:
include mock repo plugin

martijnvg committed:
removed apache license headers

martijnvg committed:
moved test to its own class. (setup logic got in the way)

martijnvg committed:
Merge remote-tracking branch 'es/master' into move_data_streams_to_xpack

martijnvg committed:
move forgotten data stream tests to xpack module

martijnvg committed:
checkstyle!

elasticmachine commented:
Pinging @elastic/es-core-features (:Core/Features/Data streams)

dakrone commented on gradle/formatting.gradle:
```diff
@@ -127,6 +127,8 @@ def projectPathsToExclude = [
   ':x-pack:plugin:ccr',
   ':x-pack:plugin:ccr:qa',
   ':x-pack:plugin:core',
+  // TODO: REMOVE BEFORE MAKING PR REVIEWABLE:
+  ':x-pack:plugin:data-streams',
```
Does this need to be removed?

dakrone commented on server/src/main/java/org/elasticsearch/cluster/metadata/MetadataIndexTemplateService.java:
```diff
@@ -1136,7 +1137,15 @@ private static void validateCompositeTemplate(final ClusterState state,
+                    final String expectedMessageIfDataStreamsIsUsedAndNotSupported =
+                        "Failed to parse mapping: Root mapping definition has unsupported parameters:  [_timestamp : {path=@timestamp}]";
+                    if (expectedMessageIfDataStreamsIsUsedAndNotSupported.equals(e.getMessage())) {
+                        // Fail like a parsing expection, since we will be moving data_stream template out of server module and
+                        // then we would fail with the same error message, like we do here.
+                        throw new XContentParseException("[index_template] unknown field [@timestamp]");
```
I think we should fail with a parsing error like `"[index_template] unknown field [data_stream]"` to make it clearer that the entire data stream configuration is unsupported, rather than just the `@timestamp` part

dakrone reviewed (approved):
LGTM, thanks Martijn! I left two pretty minor comments

martijnvg commented on gradle/formatting.gradle:
```diff
@@ -127,6 +127,8 @@ def projectPathsToExclude = [
   ':x-pack:plugin:ccr',
   ':x-pack:plugin:ccr:qa',
   ':x-pack:plugin:core',
+  // TODO: REMOVE BEFORE MAKING PR REVIEWABLE:
```
I will remove this before merging, this to make potential merge conflicts easier to resolve.

danhermann commented on x-pack/plugin/data-streams/qa/rest/src/test/resources/rest-api-spec/test/data-streams/20_unsupported_apis.yml:
```diff
@@ -0,0 +1,301 @@
+  - is_true: acknowledged
+
+---
+"Open write index for data stream opens all backing indices":
```
This is a supported API. Can move in a follow-up PR if easier.

danhermann reviewed (approved):
LGTM. Left a minor comment that can be addressed later if easier.

martijnvg committed:
moved more rest tests

martijnvg committed:
fixed a test and moved unit tests to data stream module, because these relied on _timestamp field mapper.

martijnvg committed:
fixed error message

martijnvg committed:
Merge remote-tracking branch 'es/master' into move_data_streams_to_xpack

jtibshirani reviewed (commented):
This certainly works and it's nice that we don't have to move too much code. But it feels hacky to me to rely on the timestamp metadata mapper being missing:
* The metadata mapper feels more like an implementation 'building block' than a flag that should be used to enable/ disable data streams.
* Related to this, we need to catch a low-level exception with a specific message and rethrow it.

I'll loop in the core infra team in case they have opinions, they have a lot of expertise in licensing + plugins.

elasticmachine commented:
Pinging @elastic/es-core-infra (:Core/Infra/Plugins)

talevy reviewed (commented):
I am concerned that this PR does not move all things data-streams to a plugin. Specifically, have we thought about how to handle `IndicesRequest#includeDataStreams` usage from within a plugin? Forgive my lack of context, but I do not see it mentioned, was this something that was discussed and is outside the scope of this change?

martijnvg commented:
@jtibshirani @talevy Moving the _timestamp field mapper is one of the data stream pieces that will be moved, but by moving this one first data streams we can avoid data streams being used in oss. We also move the data streams apis after this change and in a later point in time make composable index templates pluggable, so that `data_stream` definition can be moved to xpack (this is more involved, and not required to make data streams a basic licensed feature). Some data streams pieces will remain in core Elasticsearch, because moving this over requires making too many extension points in core Elasticsearch.

> The metadata mapper feels more like an implementation 'building block' than a flag that should be used to enable/ disable data streams.

It is a building block, and by moving this to xpack it also controls in what distribution data streams can be used.
I agree this isn't ideal, but it is temporary. Once more data streams bit are moved over this will not be the only
component that controls whether data streams is a basic licensed feature.

> Related to this, we need to catch a low-level exception with a specific message and rethrow it.

This will be removed once composable index templates have pluggable extensions. The exception
is thrown as if `data_streams` is not defined on composable index templates, which is what will
be thrown once the mentioned change is implemented.

martijnvg committed:
removed unused imports

martijnvg committed:
Merge remote-tracking branch 'es/master' into move_data_streams_to_xpack

martijnvg committed:
fixed unit tests

martijnvg committed:
moved full restart test

martijnvg committed:
moved data stream rolling upgrade test

martijnvg committed:
moved data stream css tests

martijnvg committed:
Merge remote-tracking branch 'es/master' into move_data_streams_to_xpack

martijnvg committed:
fixed issues after merge

martijnvg committed:
removed unneeded stuff

jtibshirani commented on server/src/main/java/org/elasticsearch/cluster/metadata/MetadataIndexTemplateService.java:
```diff
@@ -1136,7 +1137,15 @@ private static void validateCompositeTemplate(final ClusterState state,
                     }
                 } catch (Exception e) {
-                    throw new IllegalArgumentException("invalid composite mappings for [" + templateName + "]", e);
+                    final String expectedMessageIfDataStreamsIsUsedAndNotSupported =
```
Instead of relying on an exception with a specific message, what if we proactively checked if the timestamp mapper existed? We could do something like

```
boolean timestampMapperExists = tempIndexService.mapperService().isMetadataField("_timestamp");
```

and then throw if there's a `data_stream` definition but no `_timestamp` metadata mapper.

martijnvg replied:
Thanks. This is less fragile than checking the exception message.

jtibshirani commented on x-pack/plugin/data-streams/src/main/java/org/elasticsearch/xpack/datastreams/mapper/TimestampFieldMapper.java:
```diff
@@ -1,23 +1,10 @@
  */
 
-package org.elasticsearch.index.mapper;
+package org.elasticsearch.xpack.datastreams.mapper;
```
Now that we're moving this to the data streams plugin, should we update the naming to be specific to data streams? Calling it something like `_data_stream_timestamp` could also avoid confusion with the `_timestamp` metadata mapper we used to provide.

jtibshirani reviewed (commented):
> We also move the data streams apis after this change and in a later point in time make composable index templates pluggable, so that data_stream definition can be moved to xpack... Some data streams pieces will remain in core Elasticsearch, because moving this over requires making too many extension points in core Elasticsearch.

Thanks @martijnvg for the explanation, I now understand the full plan. I left some smaller review comments.

martijnvg committed:
Merge remote-tracking branch 'es/master' into move_data_streams_to_xpack

martijnvg committed:
renamed timestamp field mapper

martijnvg committed:
more robust check

martijnvg committed:
fixed tests and use feature flag

martijnvg committed:
spotless formatting

martijnvg committed:
fixed tests

martijnvg committed:
fixed rename

martijnvg committed:
mute bwc tests,
the timestamp meta field mapper is renamed from `_timestamp` to `_data_streams_timestamp`

martijnvg commented:
This PR also disables bwc tests, because the timestamp field name has been disabled.
